compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -2157,9 +2163,22 @@ static LogicalResult setConvRootConfig(mlir::FunctionOpInterface entryPointFn,
     scalableTileFlags.emplace_back(numTilingDims, false);
   }
 
+  DictionaryAttr pipelineConfig;
+  if (vecPreProcStrategy == VectorPreProcStrategy::Peeling) {
+    // Enable peeling. To this end, attach extra info to the pipeline config.
+    // This will later be extracted by LLVMCPULowerExecutableTargetPass.
+    auto ctx = convOp.getContext();
+    SmallVector<NamedAttribute> attrs;
+    auto peelAttrName = StringAttr::get(ctx, "enable_peeling");
+    auto peelAttrVal = BoolAttr::get(ctx, true);
+    attrs.push_back({peelAttrName, peelAttrVal});
+    pipelineConfig = DictionaryAttr::get(ctx, attrs);
+  }
```

Comment:
I'm confused. I thought the VectorPreProcStrategy is None for all the convolutions in `getVectorPreProcStrategy`, but we have a special logic here. Is it because we can set the strategy to peeling through CLI flags?

I can see the value about what Mahesh suggested. We decided the strategy in `KernelDispatch.cpp`, and we want to "recover" the information when adding passes to pipeline. They could conflict. It seems to be a separate issue from enabling scalable vectors for convolutions. So there are code health issue and feature support to me.

RE code health issue, I wonder if we can set the strategy on attribute in getVectorPreProcStrategy? This way we don't have floating logics everywhere to serve the need. It will need more work to make it happen cleanly, so I think it could be done in a follow-up.

RE convolution enablement: I wonder if we can retire `addConvTileAndDecomposeExpertPassPipeline`. Instead, we just use `addMultiTilingExpertPassPipeline` for convolutions as well. To achieve it, we need to add `DecomposeConvolutionToLowerDimOpsPass` to the multi-tiling pipeline, and have an additional option for `splitVectorTransfersTo`.

Reply:
Thanks for taking a look - lots of great comments 🙏🏻 

>  I thought the VectorPreProcStrategy is None for all the convolutions in getVectorPreProcStrategy, but we have a special logic here. Is it because we can set the strategy to peeling through CLI flags?

Yes. 

We have no heuristics to select a pre-processing strategy for convs. With this patch, we would allow one to be set by the user. At some point, we can also turn peeling "ON" by default for SVE. 

Btw, I always use `getVectorPreProcStrategy` to retrieve the pre-processing strategy rather than reading [clPProcStrategy](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L128-L143) directly - this way we can avoid duplicating the logic to deal with `clPProcStrategy".

> We decided the strategy in KernelDispatch.cpp, and we want to "recover" the information when adding passes to pipeline.

Yes - we need to "forward" information:
*  from [setConvRootConfig](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L2109-L2112),
*  to [LLVMCPULowerExecutableTargetPass::runOnOperation()](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L88).

I see two options for this:
1. Through a dedicated `DispatchLoweringPassPipeline` that can be used to specialise in this [switch stmt](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L114) - that's what I tried initially,
2. Through attributes - this is what the current implementation does.

I see pros and cons for both. You have much more experience with this, so I will happily follow your suggestion.

>  They could conflict

Can you expand? Not sure I know what you meant here.

> RE code health issue, I wonder if we can set the strategy on attribute in getVectorPreProcStrategy? 

I see what you mean. There's `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert`, which are virtually identical: https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L128-L139

So, if the "design" is to limit the number of lowering expert, then those two should probably be merged?

>  It will need more work to make it happen cleanly, so I think it could be done in a follow-up.

I'm happy to refactor things, but we should definitely defer that to a separate PR. I'm worried that this can lead to some surprises. I'll be under less pressure if once convs are fast for SVE :) 

>  I wonder if we can retire addConvTileAndDecomposeExpertPassPipeline. Instead, we just use addMultiTilingExpertPassPipeline for convolutions as well. 

Possibly. but looks more complex than merging `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert`:
* [addMultiTilingExpertPassPipeline](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L349-L423),
* [addConvTileAndDecomposeExpertPassPipeline](https://github.com/iree-org/iree/blob/ae64f1d4b4988f2d9dcddaaf6a21e0f61d0af411/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L425-L480)

These two have diverged a fair bit.

### Summary
There's quite a lot in here, so let me summarise. Please correct me if I got sth wrong or you disagree with the proposed approach.
1. Use **attributes** rather than a **new lowering expert** for this change,
2. In a follow-up PR, merge  `CPUDoubleTilingExpert` and `CPUDoubleTilingPeelingExpert` (this should "just work" ™️ )
3. In a follow-up PR, try merging `addConvTileAndDecomposeExpertPassPipeline` and `addMultiTilingExpertPassPipeline` (and remove `CPUConvTileAndDecomposeExpert`?). 

Basically, I'd like to land this one without 2. and 3. I'm less confident about 3., but am happy to work on both and see how it goes.

Reply:
Thanks for the good summary. It's aligned with what I'm thinking. Yes, we should go with (1) -> (2) and perhaps (3) later.